### PR TITLE
fix #158 - handle identifier list containing an indexing expression

### LINKED ISSUE #158: Parsing type expressions involving array indexing is broken
T[1].X y;

Parses/formats back into: ".X y"

The Type/Type2+TypeSuffixes  seems to assume that if it parsed "[1]" it must mean array-of-1, and then mis-handles the ".X" which follows.

There's a fundamental ambiguity here between ordinary expressions yielding types, and type-exprs involving array decls - that is not quite handled correctly by libdparse.

### PR DESCRIPTION
**Pull request by BBasile**

### CODE REVIEW

**Comment by BBasile** [comment]
The fix is wrong, `T[0].T[1].T` is well parsed but only the last indexer is stored. The problem is that the indexer cannot be parsed in `parseIdentifierOrTemplateInstance` because it is used everywhere and would have consequences  on the others constructs like the expressions.

In background the problem is that some kind of simplifications make the good fix harder than it looks.
https://dlang.org/spec/grammar.html#IdentifierList is not right in dparse.

**Comment by BBasile** [comment]
I've made a nicer fix. As expected i had to create a new AST node type, so that each indexer of the chain is stored. As expected too all the tools break: D-Scanner, D-Symbol. The user tools that used Type2 will also break.

There's no other way to do that. I repeat myself but even if the bug was trivial, libdparse simplifies the AST so that things can be re-used (`IdentifierOrTemplateChain` here). One use case was forgotten and with the fix the IdentifierOrTemplateChain would not be reusable anymore.

Finally, if this get accepted i think that we should schedule this week, two hours during which the other tools could be fixed (because it requires each a time a review, a tag, etc).

**Comment by BBasile** [comment]
This good now. Summary of the changes:

- `IdentifierList` is renamed to `DeclaratorIdentifierList`. It was used in alias declarations and in D specs, this is the name that's used: (https://dlang.org/spec/grammar.html#DeclaratorIdentifierList)
- `Type2` doesn't use an `IdentifierOrTemplateInstanceChain` but rather what's officialy an `IdentifierList` and that can include an indexer after each item of the list. This is the issue that's fixed in this PR.
- LinkageAttribute also uses this new `IdentifierList`, following the specs (https://dlang.org/spec/grammar.html#LinkageAttribute).

File: src/dparse/formatter.d [hunk 1]
```diff
@@ -3867,3 +3903,35 @@ protected:
         "postblit"
     ];
 }
+
+version (unittest)
+void testFormatNode(Node)(string sourceCode)
```
Comment (BBasile) [inline src/dparse/formatter.d]: use this new function to fix #135 and also some problems in harbored-mod (which uses dparse formatter apparently...)

File: src/dparse/parser.d [hunk 2]
```diff
@@ -3133,30 +3161,46 @@ class Parser
...
     /**
-     * Parses an IdentifierList
+     * Parses an IdentifierList.
      *
      * $(GRAMMAR $(RULEDEF identifierList):
-     *     $(LITERAL Identifier) ($(LITERAL ',') $(LITERAL Identifier))*
-     *     ;)
+     *      $(RULE identifierOrTemplateInstance)
+     *    | $(RULE identifierOrTemplateInstance) $(LITERAL '.') $(RULE identifierList)
+     *    | $(RULE identifierOrTemplateInstance) $(LITERAL '[') $(RULE assignExpression) $(LITERAL ']') $(LITERAL '.') $(RULE identifierList)
```
Comment (Hackerpilot) [inline src/dparse/parser.d]: The language spec and I disagree on the topic of "IdentifierList" being only a list of identifiers.

Comment (Hackerpilot) [inline src/dparse/parser.d]: That and the idea of "lists" being separated by commas.

Comment (BBasile) [inline src/dparse/parser.d]: It's a chain, clearly.

Comment (Hackerpilot) [inline src/dparse/parser.d]: I'm not objecting to your change. It's consistent with the spec. I'm just upset with the spec.

Comment (BBasile) [inline src/dparse/parser.d]: There must be "chain" in the name. It's not a list.

Comment (BBasile) [inline src/dparse/parser.d]: The only other use is with `extern(cpp)`.

File: src/dparse/formatter.d [hunk 3]
```diff
@@ -3867,3 +3903,35 @@ protected:
...
+
+    class CatchInterestingStuff : ASTVisitor
+    {
+        alias visit = ASTVisitor.visit;
+        override void visit(const(Node) stuff)
+        {
+            stuff.accept(this);
+            format(&fmt, stuff);
+            assert(fmt.data.canFind(code), fmt.data);
+            writeln(fmt.data);
```
Comment (BBasile) [inline src/dparse/formatter.d]: gotta remove this writeln.

**Comment by BBasile** [comment]
reverted. The changes are hard to handle.

**Pull request merged**